### ISSUE TITLE
Add `value-type-custom-property-required-list`

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#9465]
### What is the problem you're trying to solve?

I want to enforce the use of custom properties for specific types, e.g. `<color>`s. I often specify things like colours, shadows, lengths and time as custom properties. I want to enforce their use to keep my CSS code and UIs consistent.

Disallow:

```css
a { color: red; }
```

Allow: 

```css
a { color: var(--my-color); }
```

Prior art: [`stylelint-declaration-strict-value`](https://github.com/AndyOGo/stylelint-declaration-strict-value) plugin, but it has limitations, [especially for shorthand properties](https://github.com/AndyOGo/stylelint-declaration-strict-value#css-shorthand-syntax), because it's scoped to properties.

### What solution would you like to see?

A new built-in rule to enforce this common use case, especially within design systems.

- Name: `value-type-custom-property-required-list`
- Primary option: `[]` of csstree types
- Secondary options: `ignoreValues: []`
- Autofixable: No
- Message: `Expected "${value}" (<${type}>) to be a custom property`
- Description: "Specify a list of value types that require custom properties."
- Section: "Enforce conventions" -> "Allowed, disallowed & required" -> "Value"
- Config: No

We can leverage `lexer.matchProperty` if we substitute instances of `var()` (that have _known_ custom properties) with their values. 

For example, given:

```json
{
  "value-type-custom-property-required-list": [
      "color",
      "length"
    ],
}
```

Accepted:

```css
:root {
  --md: 1px;
  --fg: red;
}

a { 
  border: var(--md) solid var(--fg); 
}
```

Rejected:

```css
a { 
  border: 1px solid red;
}
```

With two problems:

```
2:11  ✖  Expected "1px" (<length>) to be a custom property 
2:21  ✖  Expected "red" (<color>) to be a custom property 
```

Thanks to `referenceFiles`, custom properties defined in other files can be known and substituted in so that the rule can work out the type of each fragment.

For example, it'd reject:

```css
a { 
  border: red solid var(--md);
}
```

With:

```
2:11  ✖  Expected "red" (<color>) to be a custom property 
```

Without substitution, csstree wouldn't be able to match because of the presence of `var()`.

I've built a PoC for the rule locally that seems to work well and is low complexity. If we're happy with the proposal, I'll open a draft PR so we can refine the edges.

---

I began thinking about this rule while exploring how to address false negatives in [`declaration-property-value-no-unknown`](https://stylelint.io/user-guide/rules/declaration-property-value-no-unknown) for values that contain `var()`. The substitution mechanism would help there, too. I've opened an issue to discuss that separately:
- https://github.com/stylelint/stylelint/issues/9464

**Comment by ybiquitous** [comment]
@jeddy3 Thanks for the proposal. This idea sounds pretty helpful. 👍🏼 

At a glance, I got the following ideas additionally:

```jsonc
{
  "value-type-custom-property-required-list": {
    "<color>": ["--fg", "--bg"], // rejects other custom properties
    "<length>": ["/^--/"] // accepts any custom properties
  }
}
```

- Map form, instead of Array, for more `var()` suggestions
- `<type>` form, instead of bare `type`, respecting [`@property`'s `syntax`](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/At-rules/@property/syntax) like `syntax: "<color>";`

What do you think?

**Comment by jeddy3** [comment]
Thank you for taking a look at the proposal!

> Map form, instead of Array

I originally designed the options like this, but felt it overlapped too much with:
- https://github.com/stylelint/stylelint/issues/9464

As both are ways of ensuring that custom properties match a type. In `declaration-property-value-no-unknown`, the check will be automatic, whereas specifying them in `value-type-custom-property-required-list` requires manual pairing (and may be redundant).

So, I simplified it to an array, i.e. require these types to be (any) custom property.

I'm not wedded to the map option, though. Shall we start with an array and revisit it during the PR review, since the tests might uncover cases where a different shape is better?

> `<type>` form

We typically use the bare form when configuring things, such as [pseudos](https://stylelint.io/user-guide/rules/selector-pseudo-class-allowed-list#arraystring) and [types](https://stylelint.io/user-guide/configure#types).

For example:

```json
{
  "languageOptions": {
    "syntax": {
      "types": { 
        "color": "| foo" 
      },
    }
  }
}
```

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdilkA2EATvCgGYQQDcWAvliADQikCWeMAckgLZyIwAPHgAd2AOjABnSY3ARMrAOYIQg4UQAuKWKSQBXPFozYU6EHiSZFepIpgB5YRpbzJZksZw4zkgJ6YNJAF3NCwvLzMNX2EYNwQ0UxNw7zkCQhCzAB8yCjNE5JQaBjCvOhMiksENGEwoSRIAbQBySSj2PBYAgFpIBRZFLtbLKCRCKCaAXWLMGmpMWV6lADEibiQNFVbfds6NCXklMW4AKxkmWGEZRE9TEC2dgIzzddiNM2mU+5gO7sX+wcCtVGUCeFmqrTMtHoNCAA).

**Comment by jeddy3** [comment]
> felt it overlapped too much

For clarity, here's the proposal for the interplay of the new and (modified) exisiting rules.

```js
export default {
  rules: {
    "declaration-property-value-no-unknown": true, // existing rule - changed scope
    "no-unknown-custom-properties": true, // existing rule - no change
    "value-no-invalid": true // new rule
    "value-type-custom-property-required-list": ["color"], // new rule
  },
};
```

```css
a { color: ); } /* Invalid value ")" (value-no-invalid) */
```

```css
a { color: 10px; } /* Unknown value "10px" for the property "color" (declaration-property-value-no-unknown) */
```

```css
a { color: red; } /* Expected "red" to be a custom property, as of type "<color>" (value-type-custom-property-required-list) */
```

```css
a { color: var(--fg); } /* Unknown custom property "var(--fg)" (no-unknown-custom-properties) */
```

```css
:root { --fg: 10px; }
a { color: var(--fg); }  /* Unknown value "10px" for the property "color", via "var(--fg)" (declaration-property-value-no-unknown) */
```

```css
:root { --fg: red; }
a { color: var(--fg); } 
```

They hopefully don't overlap while providing coverage for most mistakes.

**Comment by jeddy3** [comment]
> Map form, instead of Array

Having thought about this some more, I think the `Map` form was pointing to a use case I'd overlooked: wanting to enforce specific custom properties for a property, _not a type_. I've opened:

- https://github.com/stylelint/stylelint/issues/9470

Building on the non-overlapping examples in https://github.com/stylelint/stylelint/issues/9465#issuecomment-5464076285:

```js
export default {
  rules: {
   "declaration-property-custom-property-allowed-list": { "color": ["--fg"] }
  },
};
```

```css
:root { --fg: red; --bg: pink; }
a { color: var(--bg); }  /* Disallowed custom property "--bg" for property "color" (declaration-property-custom-property-allowed-list) */
```
